### PR TITLE
AsyncAPI

### LINKED ISSUE #1610: Add support for AsyncAPI kind
### Description

An AsyncAPI would allow users to submit single-request "jobs", and would provide an API to check the status/result of the prediction request.

Design doc: https://docs.google.com/document/d/18bN4-wBH4wF2J-M2biEVPgm_GPU2dWqV05tpca7I8zY/edit?usp=sharing

### Implementation

Would use SQS as the queue, and base autoscaling on queue length

### PR DESCRIPTION
**Pull request by miguelvr**
closes #1610 

---

checklist:

- [ ] run `make test` and `make lint`
- [ ] test manually (i.e. build/push all images, restart operator, and re-deploy APIs)
- [ ] update examples
- [ ] update docs and add any new files to `summary.md` (view in [gitbook](https://cortex-labs.gitbook.io/staging/-MOmCGMADSRNQahK3Kox/) after merging)
- [ ] cherry-pick into release branches if applicable
- [ ] alert the dev team if the dev environment changed

### CODE REVIEW

File: pkg/cortex/serve/cortex_internal/lib/api/async.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+                            f"multiple definitions for {target_class_name} class found; please check "
+                            f"your imports and class definitions and ensure that there is only one "
+                            f"predictor class definition"
+                        )
+                    predictor_class = class_df[1]
+            if predictor_class is None:
+                raise UserException(f"{target_class_name} class is not defined")
+            return predictor_class
+        else:
+            raise UserException("no callable class or function were provided")
```
Comment (miguelvr) [inline pkg/cortex/serve/cortex_internal/lib/api/async.py:174]: update error message

**Review by RobertLucian — APPROVED** [review]

**Comment by lminer** [comment]
This is great! A couple of questions. Can this be used with the TensorFlow predictor (i.e. will `self.client.predict` return a job id)? Will docs for this be released soon?

**Comment by vishalbollu** [comment]
AsyncAPI is currently not compatible with Tensorflow predictor type in this release, only the Python predictor type is. However we are planning to iterate on AsyncAPI further and will keep you posted. Here are the [docs](https://docs.cortex.dev/workloads/introduction) for async.

**Comment by lminer** [comment]
Thanks for the update @vishalbollu. Is there an issue for the Tensorflow predictor type that I could follow?

**Comment by vishalbollu** [comment]
I just created one https://github.com/cortexlabs/cortex/issues/1980. This will be updated accordingly.

**Pull request merged**